File: android-java/app/src/main/java/com/example/dittotasks/TasksApplication.java
package com.example.dittotasks;

import android.app.Application;
import android.util.Log;

/**
 * Creates the process-global Ditto instance exactly once, at app startup, and starts
 * sync. Ditto lives for the whole process and is never tied to an Activity's lifecycle,
 * so a configuration change such as rotation recreates the Activity without recreating
 * Ditto (which would contend on the persistence-directory lock).
 */
public class TasksApplication extends Application {
    private static final String TAG = "TasksApplication";

    @Override
    public void onCreate() {
        super.onCreate();
        try {
            DittoManager.initialize(
                    BuildConfig.DITTO_DATABASE_ID,
                    BuildConfig.DITTO_SERVER_URL,
                    BuildConfig.DITTO_DEVELOPMENT_TOKEN,
                    BuildConfig.DITTO_OFFLINE_LICENSE_TOKEN);
            DittoManager.startSync();
        } catch (Throwable ex) {
            Log.e(TAG, "Failed to initialize Ditto", ex);
            throw ex;
        }
    }
}

File: android-kotlin/QuickStartTasks/app/src/main/java/live/ditto/quickstart/tasks/TasksApplication.kt
package live.ditto.quickstart.tasks

import android.app.Application
import android.content.Context
import com.ditto.kotlin.DittoConfig
import com.ditto.kotlin.DittoLog

class TasksApplication : Application() {

    private val tag = "TaskApplication"

    companion object {
        private var instance: TasksApplication? = null

        fun applicationContext(): Context {
            return instance!!.applicationContext
        }
    }

    init {
        instance = this
    }

    override fun onCreate() {
        super.onCreate()
        initializeDitto()
    }

    private fun initializeDitto() {
        try {
            val config = DittoConfig(
                databaseId = BuildConfig.DITTO_DATABASE_ID,
                connect = DittoConfig.Connect.Server(url = BuildConfig.DITTO_SERVER_URL)
            )

            DittoManager.initialize(config, BuildConfig.DITTO_DEVELOPMENT_TOKEN)
        } catch (ex: Throwable) {
            DittoLog.e(tag, "Failed to initialize Ditto: $ex")
            throw ex
        }
    }
}
